contribute/design/1.x/cstor/pdb/20200113-cstor-poddisruptionbudget.md

```diff
@@ -0,0 +1,191 @@
+controller. CVC controller will schedule the replicas among the pools after that
+CVC controller will create a PDB among the pool pods where the volume replicas
+are scheduled. PDB will be created only if PDB among those pools is not yet available.
+If PDB already exists, a label will be added on CVC with the PDB name.
+
+Example:
+- Provisioned a CSPC with five pool specs that intern creates 5 cStor Pools.
+- Now, when 'n' no.of HA volumes are provisioned(with replica count as 3) then
+  volume replicas can be schedule in any available cStor pools.
+- After volume replicas are placed then cStorVolumeConfig controller will
+  create PDB among those cStor pools.
+- After successful creation of PDB corresponding cStorVolumeConfig will be
+  updated with PDB label.
+
+So when `n` no.of HA volumes are provisioned then sigma(nCr)[where n is no.of pools,
+r is no.of replicas] no.of PDB's will be created. For example if `n` pools are
+created then at max nC3 + nC4 + nC5 PDB's will be created(Where 3, 4 & 5 are
+replicas).
+
+Note:
+    - Further optimization is done on PDB creation. PDB creation can be avoided when
+      the current volume replica pools are already subset of any existing PDB then
+      current CVC will refer to existing superset PDB(instead of creating new PDB).
+      But not vice versa.
+      Example:
+         If PDB2 needs to be created among pool1, pool2 and pool3 this can be
+         avoided if there is any existing PDB created among pool1, pool2, pool3 and pool4.
+    - PDB will be created only for HA volumes(i.e only for the volumes greater
+      than or equal to 3 replica count).
+    - During deprovision of volume PDB will be deleted automatically if no
+      cStorVolume refers to PDB.
+
+### Steps to perform user stories
+- User has to create HighAvailable volumes on cStor then PDB will be created by
+  default.
+
+### Low Level Design
+
+#### Work Flow
+- User/Admin provisioned HA volume, Now CVC controller schedules the replicas in
+  the cStor pools. After scheduling the replicas CVC controller will try to get
+  PDB that was created among those cStor pools. If PDB doesn't exist then PDB will
+  be created with following details:
+    - All the pool names where the volume replicas are scheduled will be added as
+      labels on PDB[Why? to identify the PDB without iterating over all the existing PDBS].
+- Once PDB exists for those pools, PDB name will be added as label(openebs.io/pod-disruption-budget: <pdb_name>)
+  on CVC[why? To identify how many volumes are referring particular PDB. So
+  during deprovisioning time if there are no reference to PDB then PDB can be deleted].
+- During deprovision of volume CVC controller will verify is there any other cStor
+  volume referring to this PDB(i.e PDB created among current volume replica pools).
+  If no such volume exists then PDB will be destroyed and then finalizers will be
+  removed from CVC or else finalizers will be removed from CVC.
```

Comment:
removed from CVC or else finalizers will be removed from CVC and trigger the delete.

Reply:
We already mentioned up on deprovisining of volume i.e deletion is already triggered.